c/collections.c

```diff
@@ -1417,6 +1417,84 @@ void *qRemove(Queue *q){
 #endif /* END OF OS-VARIANT Queue stuff */
 
 
+/* The Array List (flexible Array thing that smells like Java and javascript) */
+
+static void *arrayListAlloc(ArrayList *list, uint32_t size, char *location){
+  if (list->slh){
+    return (void*)SLHAlloc(list->slh,size);
+  } else{
+    return safeMalloc(size,location);
+  }
+}
+
+ArrayList *makeArrayList(){
+  ArrayList *list = (ArrayList*)safeMalloc(sizeof(ArrayList),"ArrayList");
+  list->capacity = 8;
+  list->size     = 0;
+  list->array = (void**)safeMalloc(list->capacity*sizeof(void*),"ArrayListArray");
+  list->slh = NULL;
+  return list;
+}
+
+void arrayListFree(ArrayList *list){
+  if (list->slh == NULL){
+    safeFree((char*)list->array,list->capacity*sizeof(void*));
+    safeFree((char*)list,sizeof(ArrayList));
+  }
+}
+
+void initEmbeddedArrayList(ArrayList *list,
+			   ShortLivedHeap *slh){ /* can be null for use safeMalloc - standard heap alloc */
+  list->capacity = 8;
+  list->size     = 0;
+  list->slh      = slh;
+  list->array = (void**)arrayListAlloc(list,list->capacity*sizeof(void*),"ArrayListArray");  
+}
+
+void arrayListAdd(ArrayList *list, void *thing){
```

Comment:
Could we use `const void*`? That way we can pass const value without the compiler producing warnings/errors.

Reply:
What's an example of where this gets a warning?

Reply:
I believe this one would get a warning message:
```c
const char* str1 = "hello";
arrayListAdd(myList, str1);
```